Memotest

maurperez opened the pull request.

Commit:
Subiendo imagenes de frutas

Commit:
Desarrolo de funciones

El juego ya esta casi terminado, bloquear el juego cuando no se le dio a comenzar  o cuando esta comparando cartas, desarrollar un contador de segundos y correcciones menores

Commit:
Eliminando archivo creado para debuggear

Commit:
Commit para correcion

No hubo avances

Commit:
Avences en cuanto a voltear cartas

Volteo de cartas completado, falta unicamente agregarlo el efecto flip.
Estamos en proceso de desarrolo del conntador

Commit:
Desarrollo del contador

Commit:
Grandes optimizaciones

Hubo grandes cambios en realacion al componente CuadroMemotest, ya que ahora se utiliza un componente nuevo que ademas de hacer mas eficiente la carta, agrega un efecto Flip cuando se clickea
Gracias a lo anterior se hizo mucho mas simple y eficiente el resto del codigo

Commit:
Juego finalizado

asus753 commented:
![memotest-example](https://user-images.githubusercontent.com/56839325/78610066-72f02b00-783a-11ea-9468-44d44288175a.png)

jabsatz commented on src/components/pages/memotest.js:
```diff
@@ -0,0 +1,219 @@
+import React, {useState, useRef, useEffect} from 'react'
+import PropTypes from 'prop-types';
+import './memotest.css'
+import ReactCardFlip from 'react-card-flip';
```
Me parece excelente que ya tengas el concepto de importar otras librerías para utilidades así :)

jabsatz commented on src/components/pages/memotest.js:
```diff
@@ -0,0 +1,219 @@
+			return Math.random()-0.5
+	})
+	
+	const frutas = []
+	desordenado.forEach(e => frutas.push({
+		src : require(`./imagenes-memotest/${e}.jpg`),
+		estado : false
+	}))
```
esto lo podés transformar en un `map` si querés

```
const frutas = desordenado.map(e => {
  return {
		src : require(`./imagenes-memotest/${e}.jpg`),
		estado : false
	}
})
```

jabsatz replied:
el require de acá lo podés dejar, te vas a volver loco para hacer que esto funcione con imports comunes

jabsatz commented on src/components/pages/memotest.js:
```diff
@@ -0,0 +1,219 @@
+
+const useTableroCartas = () => {
+	const [frutas, setFrutas] = useState(crearObjetoFrutas(arrayOrdenadoDeFrutas))
+	const frutasRef = useRef(frutas)
+	frutasRef.current = frutas
```
esta ref no hace falta, podés directamente usar el estado de `frutas`, que siempre va a valer lo mismo.

abajo donde hacés `[...frutasRef.current]` si lo reemplazás con `[...frutas]` anda igual y tenés menos complejidad en el hook

maurperez replied:
Ahhh mira funciona, al inicio del proyecto había hecho algo medio raro en que si no usaba el Ref no funcionaba correctamente y después di por sentado que era necesario :thumbsup: :thumbsup:

jabsatz commented on src/components/pages/memotest.js:
```diff
@@ -0,0 +1,219 @@
+	const evaluarJuegoFinalizado = ()=>{
+		const todasVolteadas = frutas.every((carta) => carta.estado === true)
+
+		if(todasVolteadas && contador>0 && juegoActivo){
+			setJuegoActivo(false)
+			setJuegoFinalizado('gano')
+		}else if(!todasVolteadas && contador === 0 && juegoActivo){
+			setJuegoActivo(false)
+			setJuegoFinalizado('perdio')
+		}
```
Los estados de `juegoActivo` y `juegoFinalizado` parece que siempre se setean juntos, podrías usar un solo estado como `juegoEtapa` o `juegoEstado` que tenga valores de `['no comenzado', 'activo', 'gano', 'perdio']`, o algo así.
Eso te simplifica bastante el estado también.

jabsatz commented on src/components/pages/memotest.js:
```diff
@@ -0,0 +1,219 @@
+		}
+	}
+
+	const evaluarCartasVolteadas = () => {
+		if(frutasClickeadas.length === 2){
+			if(frutasClickeadas[0].src === frutasClickeadas[1].src){
+				const indice1 = frutasClickeadas[0].indice, indice2 = frutasClickeadas[1].indice
+				setTimeout(()=>{
+					quitarCartas(indice1, indice2)
+				},700)
+				frutasClickeadas.splice(0,2)
+			}else{
+				const indice1 = frutasClickeadas[0].indice, indice2 = frutasClickeadas[1].indice
+				setTimeout(()=>{
+					ocultarCartas(indice1, indice2)
+				},700)
+				frutasClickeadas.splice(0,2)
+		}}
+		
+	}
```
Entiendo lo que querés hacer acá, pero me parece que `frutasClickeadas` es una variable que te podría complicar mucho. Como es una variable por separado del state `frutas`, siempre que modificás ese state vas a tener que modificar esta variable también, sin mencionar que al estar por afuera de este hook podrías modificar esa variable desde otro lado.

Lo que haría yo acá es declarar `frutasClickeadas` dentro de esta función, así:
```
let frutasClickeadas = [];
frutas.forEach((fruta, indice) => {
  if(fruta.estado) {
    frutasClickeadas.push({src: nuevoFrutas[index].src, indice : index})
  }
})
```

De esta manera podés estar seguro que frutasClickeadas no va a cambiar inesperadamente: nace en este momento a partir de tu estado y cuando termina la función, muere.

maurperez replied:
Disculpa que te responda ahora recién, el problema que tuve al intentar hacerlo así como vos dijiste es que se buguea antes clicks demasiados rápidos, ej si llegas a dar vuelta 5 cartas antes de los 700ms del setTimeOut el juego se rompe. Ademas no habría forma de saber cual fue clickeada antes de cual.
La solución que encontré sin tener que usar una variable global es crear frutasClickeadas con un useState dentro del custom Hook. Lo estuve probando y funciona. Si podes responderme que tal te parece la solución. Ahora hago un commit con la solución

jabsatz commented on src/components/pages/memotest.js:
```diff
@@ -0,0 +1,219 @@
+	const quitarCartas = (indice1, indice2) => {
+		const newFrutas = [...frutasRef.current]
+
+		newFrutas[indice1].src = require('./imagenes-memotest/fondo.jpg')
+		newFrutas[indice2].src = require('./imagenes-memotest/fondo.jpg')
```
esto funciona? creería que no podés usar `require` acá

maurperez replied:
No tuve ningún problema al respecto, no sabia bien como importar las imágenes y vi en stack que se podía hacer así. Como seria la forma correcta?

jabsatz replied:
la forma correcta sería, al principio del archivo hacer algo así como
`import imagenFondo from './imagenes-memotest/fondo.jpg'`
y después acá hacer
`newFrutas[indice1].src = imagenFondo`

Lo que me hace ruido es que `require` es la forma de importar cosas que usa nodeJS, y esto es codigo que compila al browser, por lo que no debería permitir estas cosas. Seguramente create-react-app esté compilando esto a un import común o algo así, pero en muchos otros ambientes un código así no funcionaría.

(Esto te lo marco ahora pero si no viste nada de node puede que no lo entiendas, no pasa nada igual)

maurperez replied:
No, no tengo nada de idea de nodeJS. 

> `import imagenFondo from './imagenes-memotest/fondo.jpg'`

Esa forma también la había visto pero me pareció medio raro escribir 9 imports, uno por cada imagen entonces use el `require`. Pero ahora lo cambio, yo pensaba que era algo de JS nomas :sweat_smile:

jabsatz reviewed (commented):
muy bien che, te dejo unos comments 👍

Commit:
Correciones

cartasClickeadas sujeto a revision

Commit:
Correcion error

Si todas la cartas se clickeaban en un plazo menor a 700 ms el juego se daba por ganada. Se corrigio agragando una propiedad especifica al objeto de cada fruta.

jabsatz commented on src/components/pages/memotest.js:
```diff
@@ -0,0 +1,221 @@
+	const [frutas, setFrutas] = useState(crearObjetoFrutas(arrayOrdenadoDeFrutas))
+	const [contador, setContador] = useState(60)
+	const [juegoEstado, setJuegoEstado] = useState('inactivo')
+	const [cartasClickeadas, setCartas] = useState([])
```
lo único que te recomendaría acá es que el setter se llame `setCartasClickeadas` así tiene el mismo nombre que el estado y se entiende más cuando lo usás que van de la mano

maurperez replied:
Muchísimas gracias por las correcciones juli!

jabsatz reviewed (approved):
10 puntos!

Commit:
correciones menores
